More robust config validation logic

conorsch opened the pull request:
Retains the two previous checks:

  1. Does config.json exist?
  2. Does sd-journalist.sec exist?

Adds several checks verifying the contents of config.json.
There's a minimum of user-friendly messages explaining specifically what
the problem is. We can add those messages over time, aiming to balance
utility for Admins and developers.

Assumes that a v2 Onion URL is used. We can update that logic when we
add support for v3 Onion URLs.

Closes #240.

### Testing

- [ ] `make validate` passes in dom0 with v3 Onion URL info
- [ ] `make validate` passes in dom0 with v2 Onion URL info
- [ ] `make validate` fails when a known-bad value is in `config.json` as appending characters to the Onion URL

rmol commented on scripts/validate-config:
```diff
@@ -0,0 +1,72 @@
+        assert usb_config.startswith("sys-usb")
+
+    def confirm_submission_privkey_file(self):
+        assert os.path.exists("sd-journalist.sec")
```
How about running  `gpg sd-journalist.sec` to validate its content?

rmol commented on scripts/validate-config:
```diff
@@ -0,0 +1,72 @@
+        assert "key" in self.config["hidserv"]
+        assert re.match(TOR_V2_AUTH_COOKIE_REGEX, self.config["hidserv"]["key"])
+
+    def confirm_usb_export_device(self):
```
This is good but won't be necessary after #324 lands.

conorsch commented:
@rmol Great comments, thanks for taking a look. I've rebased and appended a few changes here, namely to add v3 support, and implement the `gpg` call you suggested. Also updated the test plan to ensure that both v2 & v3 config files work as expected.

> If this is something that will eventually be reporting to journalists, we might want to print nicer messages. If it's just for us, asserts are OK.

Very much dev-focused right now. Deferring a friendlier experience for when we tackle the overall admin provisioning story. At that time, we'll likely crib a fair amount of code from https://github.com/freedomofpress/securedrop/blob/43245829ff93e0a84999526f7ecd0435f89eb52a/admin/securedrop_admin/__init__.py#L60

Ready for re-review!

rmol commented on scripts/validate-config:
```diff
@@ -0,0 +1,100 @@
+        assert "usb" in self.config
+        assert "device" in self.config["usb"]
+        usb_config = self.config["usb"]["device"]
+        assert usb_config.startswith("sys-usb")
```
I realized this time through that this doesn't check for a completely valid Qubes USB identifier; it should technically be another regex check for `r'^sys-usb:\d+-\d+$'`. I'm inclined to let it go at this point, given its impending obsolescence, but could also make that change if anyone feels differently.

rmol replied:
Yeah, going to merge this as-is and just get the auto attachment in ASAP.
